### PR TITLE
JS: Refactorization of NoSQL.qll to use API-graphs

### PR DESCRIPTION
**Pull request by erik-krogh**
~~A bit of precision has been lost compared to the previous implementation, due to limitations of the API-graphs API. 
But interestingly, no test case failed as a result.  
The imprecisions are in the top of NoSQL.qll (`getAMongoClient()` / `getAMongoDb()`).~~

I added a new convenience method to `API::EntryPoint`, which allows me to easily get an `API::Node` from an `EntryPoint`. 

I had to change the implementation in some places to avoid API-graph related gotchas.  
For example: instead of `Mongoose::InvokeNode` (modelling a call to a Mongoose function) I now have a `Mongoose::MongooseFunction` (modelling the Mongoose function itself).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll
Comment (max-schaefer) [inline javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:33]: Can't you use the newly introduced `getAnImmediateUse` to rule this out? The immediate use would be the `DataFlow::Parameter` itself, so you can check its name.

**Comment by erik-krogh** [comment]
An evaluation came back bad because of a bad join-order (see below tuple counts). 

I changed the predicate a bit such that the performance should be good now. 

```
[2020-10-03 18:44:59] (458s) Tuple counts for NoSQL::Mongoose::Document::getAMongooseDocument#f#join_rhs#1:
1         ~0%     {4} r1 = CONSTANT(unique string, unique string, unique string, unique int)["then","0","parameter 0",0]
1         ~0%     {4} r2 = SELECT r1 ON r1.<3> >= -1
1         ~0%     {4} r3 = SELECT r2 ON 0 >= -1
1         ~0%     {2} r4 = SCAN r3 OUTPUT r3.<2>, ("member " ++ r3.<0>)
1         ~0%     {4} r5 = CONSTANT(unique string, unique string, unique string, unique int)["exec","1","parameter 1",1]
1         ~0%     {4} r6 = SELECT r5 ON r5.<3> >= -1
1         ~0%     {4} r7 = SELECT r6 ON 0 >= -1
1         ~0%     {2} r8 = SCAN r7 OUTPUT r7.<2>, ("member " ++ r7.<0>)
2         ~0%     {2} r9 = r4 \/ r8
272334    ~4%     {3} r10 = JOIN r9 WITH ApiGraphs::API::Impl::edge#2#fff_102#join_rhs AS R ON FIRST 1 OUTPUT r9.<1>, R.<1>, R.<2>
548812365 ~0%     {4} r11 = JOIN r10 WITH ApiGraphs::API::Impl::edge#2#fff_102#join_rhs AS R ON FIRST 1 OUTPUT R.<1>, r10.<1>, r10.<2>, R.<2>
                  return r11
```

**Comment by erik-krogh** [comment]
[A new evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1601818775641) came back, and it's looking good. 

I think it might actually be a performance improvement.  
(And no changes to `--metric TaintSinks.ql`, which is a good).

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll
Comment (max-schaefer) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:60]: To expand on the above: you are meant to be able to do `root().getMember("mongodb").getMember("Collection").getInstance()` (cf [this test](https://github.com/github/codeql/blob/d38121f54a5f6e9144a99b6beebf9562a5969cdd/javascript/ql/test/ApiGraphs/typed/index.ts#L14)). If that doesn't work, I'll fix it.

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll
Comment (max-schaefer) [inline javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:240]:
```suggestion
     * Gets a API-graph node referring to a Mongoose Model object.
```

**Review by max-schaefer — CHANGES_REQUESTED** [review]
Fantastic work, thanks! I have a few suggestions for improving terminological consistency, and one TypeScript-related question. I'm fine with addressing the latter in a follow-up PR to avoid upsetting the performance apple cart.

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll:60]: > To expand on the above: you are meant to be able to do `root().getMember("mongodb").getMember("Collection").getInstance()` (cf [this test](https://github.com/github/codeql/blob/d38121f54a5f6e9144a99b6beebf9562a5969cdd/javascript/ql/test/ApiGraphs/typed/index.ts#L14)). If that doesn't work, I'll fix it.

That worked, except it ends up looking like: 
```CodeQL
result = API::root().getASuccessor(API::EdgeLabel::mod("mongodb")).getMember("Collection").getInstance()
```

The `mongodb` type has edge `"mod mongodb"`, but it is not a module-import so using `API::moduleImport("mongodb")` does not work. 

The API could use a cleanup, but lets do that later.

Comment (asgerf) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll]: FWIW I ran into the same issue with the Angular2 model and held off on using API graphs for this reason.

The snippet you posted leaves more questions than it answers. Is `root().getMember(x)` the same as `API::moduleImport(x)`? If `Collection` is the name of an interface, what exactly does `getInstance` mean?

Comment (max-schaefer) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll]: > That worked, except it ends up looking like: (...)

:see_no_evil:

> The mongodb type has edge "mod mongodb", but it is not a module-import so using API::moduleImport("mongodb") does not work.

That's an oversight. I'll fix it, then we can see about using it in the NoSQL model.

> If Collection is the name of an interface, what exactly does getInstance mean?

Ah, I had hoped that the concept of being an instance of a type would be natural, but if it's not we can introduce a new kind of edge for it. Any ideas for what to call it?

**Comment by erik-krogh** [comment]
I think I've found a potential performance improvement. But that can wait for later, this PR is good to merge as is (when it comes to performance). 

@max-schaefer 
Have you considered making a `newtype Label` in API-Graphs? 
Currently the `ApiGraphs::API::Node::getAMember` predicate performs quite badly on `vscode`, and the reason looks to be the `string` labels ([tuple counts and clause timings](https://gist.github.com/erik-krogh/d5c35a0c692401f24b3feeeac8141245)). 

I did a quick-and-dirty experiment trying to make a `newtype Label`, and `vscode` became significantly faster.    
([See clause timings in notes](https://gist.github.com/erik-krogh/d5c35a0c692401f24b3feeeac8141245)).

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll]: > Ah, I had hoped that the concept of being an instance of a type would be natural, but if it's not we can introduce a new kind of edge for it. Any ideas for what to call it?

I think `getInstance()` is natural. We just need some utility-predicate to work with `CanonicalName`s.

**Comment by max-schaefer** [comment]
> Have you considered making a newtype Label in API-Graphs?

Yes. It's something I've gone back and forth about a couple of times. The problem with the `newtype` approach is that it requires us to specify at the definition site which labels we want, which is a little error-prone. With the `bindingset` annotation we can get away with doing that at the use site, which is more convenient. I experimented with having an abstract class of label names that is used in the definition and extended at the use sites, but that became so verbose that I went back to `bindingset`s.

Definitely worth revisiting that decision if it benefits performance.

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll
Comment (asgerf) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll]: I'd be in favor of a classless predicate that's a direct replacement for `hasUnderlyingType`:
```codeql
API::nodeOfType("mongodb", "Collection")
```

I think representing values, type names, and type-namespaces as the same kind of node will lead to confusion, as most of the `API::Node` predicates are meaningless for type names and type-namespaces. Going straight to the "value" space means you can never mix up the two kinds of nodes.

I don't think we need to change the underlying encoding.

**Comment by erik-krogh** [comment]
> Definitely worth revisiting that decision if it benefits performance.

This is the implementation I used in my small experiment:  https://github.com/erik-krogh/ql/commit/0ec5e0e3428d3bb57ced431a6addae66242ff952 

(I forgot to add the link in the previous comment).

**Comment by max-schaefer** [comment]
Thanks for sharing. I think that implementation nicely highlights the duplication you get with that approach.

File: javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll
Comment (max-schaefer) [reply javascript/ql/src/semmle/javascript/frameworks/NoSQL.qll]: Something like [this](https://github.com/max-schaefer/ql/tree/js/api-graphs-type-handling-improvements)?

**Comment by max-schaefer** [comment]
I'd prefer to remove the last commit. Then we have something that's identical (modulo comments) to the version that was evaluated. I'll fix the TypeScript API in a separate PR.

**Comment by erik-krogh** [comment]
> I'd prefer to remove the last commit. Then we have something that's identical (modulo comments) to the version that was evaluated. I'll fix the TypeScript API in a separate PR.

I've reverted the last commit.

**Review by max-schaefer — APPROVED** [review]

**Pull request merged**
